usr/src/uts/common/fs/zfs/zap.c

```diff
@@ -505,6 +505,16 @@ zap_get_leaf_byblk(zap_t *zap, uint64_t blkid, dmu_tx_t *tx, krw_t lt,
 
 	ASSERT(RW_LOCK_HELD(&zap->zap_rwlock));
 
+	/*
+	 * If system crashed just after dmu_free_long_range in zfs_rmnode, we
+	 * would be left with an empty xattr dir in delete queue. blkid=0
+	 * would be passed in when doing zfs_purgedir. If that's the case we
+	 * should just return immediately. The underlying objects should
+	 * already be freed, so this should be perfectly fine.
+	 */
+	if (blkid == 0)
```

Comment:
I think this is the wrong way to address the problem of a corrupt (zero-filled) ZAP object.  We should be explicitly detecting an operation on an empty zapobj as soon as possible, e.g. in zap_lockdir, we can check that zap_block_type is ZBT_HEADER or ZBT_MICRO, and zap_magic is ZAP_MAGIC.

We also need to address the root cause of this corruption, which I think you've identified: when zfs_rmnode() is called on the xattr dir, we do zfs_purgedir() and then dmu_free_long_range().  If we crash after that, we will do the zfs_purgedir() again, on a now-corrupt ZAP.  The simplest solution would be to not call dmu_free_long_range() on XATTR dir's.  Given that the typical number of attributes is very small, freeing the whole object in one txg should be OK.